File: massmailing/src/main/java/fr/openent/massmailing/service/SettingsService.java
package fr.openent.massmailing.service;

import fr.openent.massmailing.enums.MailingType;
import fr.openent.massmailing.model.Mailing.*;
import fr.wseduc.webutils.Either;
import io.vertx.core.*;
import io.vertx.core.json.JsonObject;

import java.util.List;

public interface SettingsService {

    /**
     * Retrieve all templates for given type and given structure identifier.
     *
     * @param type      Mailing type
     * @param structure Structure identifier
     * @param listCategory list of category
     */
    Future<List<Template>> getTemplates(MailingType type, String structure, List<String> listCategory);

    /**
     * Create given template.  Must contains structure identifier, template name, template content et template type
     *
     * @param template Template
     * @param userId   User identifier
     * @param handler  Function handler returning data
     */
    void createTemplate(JsonObject template, String userId, Handler<Either<String, JsonObject>> handler);

    /**
     * Update given template
     *
     * @param id       Template identifier
     * @param template Template object. Must contains structure identifier, template name, template content et template type
     * @param handler  Function handler returning data
     */
    void updateTemplate(Integer id, JsonObject template, Handler<Either<String, JsonObject>> handler);

    /**
     * Delete given template
     *
     * @param id      Template identifier
     * @param handler Function handler returning data
     */
    void deleteTemplate(Integer id, Handler<Either<String, JsonObject>> handler);

    /**
     * Retrieve template based on template identifier and structure identifier
     *
     * @param id        Template identifier
     * @param structure Structure identifier
     * @param handler   Function handler returning data
     */
    void get(MailingType type, Integer id, String structure, Handler<Either<String, JsonObject>> handler);
}

File: massmailing/src/main/resources/public/ts/services/SettingsService.ts
import {ng} from 'entcore';
import http from 'axios';
import {MailTemplateCategory} from "@common/core/enum/mail-template-category";

export interface Template {
    id?: number
    structure_id?: string
    name: string,
    content: string,
    type?: 'MAIL' | 'PDF' | 'SMS',
    category: MailTemplateCategory
}

export interface SettingsService {
    /**
     * Retrieve templates based on given type and structure identifier
     * @param type template type. Should be MAIL, PDF or SMS
     * @param structure structure identifier
     * @param category category identifier
     */
    get(type: 'MAIL' | 'PDF' | 'SMS', structure: string, category: string): Promise<any>;

    /**
     * Create given template
     * @param template template to create
     */
    create(template: Template): Promise<any>;

    /**
     * Update given template
     * @param template template to update
     */
    update(template: Template): Promise<any>;

    /**
     * Delete given template
     * @param template template to delete
     */
    delete(template: Template): Promise<any>;
}

export const settingsService: SettingsService = {
    get: async function (type: "MAIL" | "PDF" | "SMS", structure: string, category: string): Promise<any> {
        try {
            const {data} = await http.get(`/massmailing/settings/templates/${type}?structure=${structure}&category=${category}`);
            return data;
        } catch (e) {
            throw e;
        }
    },

    create: async function (template: Template): Promise<any> {
        try {
            const {data} = await http.post(`/massmailing/settings/templates`, template);
            return data;
        } catch (e) {
            throw e;
        }
    },
    update: async function (template: Template): Promise<any> {
        try {
            const {data} = await http.put(`/massmailing/settings/templates/${template.id}`, template);
            return data;
        } catch (e) {
            throw e;
        }
    },
    delete: async function (template: Template): Promise<any> {
        try {
            const {data} = await http.delete(`/massmailing/settings/templates/${template.id}`);
            return data;
        } catch (e) {
            throw e;
        }
    }
};

export const ngSettingsService = ng.service('SettingsService', (): SettingsService => settingsService);